### ISSUE TITLE
Inlining `Kern` and subclasses.

### ISSUE THREAD

**Issue by LonelyCat124** [stfc/PSyclone#2216]
Here is a "standalone" test that fails on `master` to do with inlining with `GOKern`s (it can be all just added to the end of the inline transformation test file and will fail).

```
GOCEAN_BASE_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                os.pardir, os.pardir, "test_files",
                                "gocean1p0")
from psyclone.parse.algorithm import parse
from psyclone.psyGen import PSyFactory, CodedKern
from psyclone.domain.common.transformations import KernelModuleInlineTrans

def test_inline_failure():
    ''' Test the apply method of the InlineTrans with gocean. '''
    _, invoke_info = parse(os.path.join(GOCEAN_BASE_PATH, "single_invoke.f90"),
                           api="gocean1.0")
    inline_trans = InlineTrans()
    kintrans = KernelModuleInlineTrans()
    psy = PSyFactory("gocean1.0", distributed_memory=False).\
        create(invoke_info)
    schedule = psy.invokes.invoke_list[0].schedule

    calls = schedule.walk(CodedKern)
    kintrans.apply(calls[0])
    calls[0].lower_to_language_level()
    calls = schedule.walk(Call)
    inline_trans.apply(calls[0])

    code = str(psy.gen)
```
To be able to apply the inline transformation here, I need to get the `GOKern` and convert it into a `Call` (which is usually done at lowering, but I force it to happen early both here, and inside the omp task transformation apply function).

The test breaks at the end of `_find_routine` as the `routine_symbol` is none of `is_modulevar`, `is_unresolved`, nor `is_import`. In fact, if I look at `call_node.root.children[:]` it turns out the routine is `call_node.root.children[1]`, so in the current scope as a `resolved` symbol, however this is not searched and just falls through to the `InternalError`.

I'm not sure if its sensible to search the current scope for any `Routine` and if it is the appropriate routine, to return that routine? Notably in the future we need to check the scope is correct (e.g. for a Routine contained inside a Routine that does not contain the `call_node` we need to not find it and still fall through to the error).

Does this seem like a reasonable idea or am I just using `InlineTrans` badly here?

@arporter @sergisio

**Comment by sergisiso** [comment]
There is the  src/psyclone/tests/psyir/transformations/inline_trans_test.py::test_apply_gocean_kern that test the same as the provided example and sets up the include path. The test xfails with #924, but I think its better to keep this new issue as that one is a different restriction.

**Comment by arporter** [comment]
Related to this, PSyclone will currently refuse to inline any call where any of the actual arguments are of `UnknownType`. Now that `UnknownFortranType` can have `partial_datatype` information we can do better (and in fact need to for LFRic where all field/operator arguments will be of UnknownFortranType because they have the pointer attribute).

**Comment by arporter** [comment]
I've created branch `2216_inlining_with_partial_type` to keep what I have so far for this.

**Comment by LonelyCat124** [comment]
I added `-I /path/to/dl_esm_inf/src` and I'm still hitting the same issue.
Looking in pdb I see I have a `DeferredType` (not an `UnknownType`) - does this mean I should be able to force PSyclone to work this out before here - or is this just the same issue you've been working at in #924 ?

**Comment by arporter** [comment]
Have you tried the `resolve_imports` suggested by Sergi? My work in #924 is to support the NEMO API but if you're in GOcean you should be OK.

**Comment by LonelyCat124** [comment]
I hadn't before - I think I am still hitting the same issue Sergi highlighted before with subcomponents.
As far as I can work out the `ssha_t%data` StructureReference's datatype is not being resolved and using `resolve_imports` does appear to resolve all the top level imporst but not the lower level ones.

Asking the symbol table it seems to know that `ssha_t` is a `DataSymbol<r2d_field: DataTypeSymbol, Argument(Access.READWRITE)>`, however the `%data` access still results in a `DeferredType`.

However if i dig a bit more, even after doing `schedule.symbol_table.resolve_imports()` if I check if PSyclone knows what the `r2d_field` (`print(schedule.symbol_table.get_symbols()['ssha_t'].datatype.datatype)`) is I still get `DeferredType` - I'm not sure if this is resolvable? I've tried `resolved = schedule.symbol_table.get_symbols()['ssha_t'].datatype.resolve_deferred()` but `resolved.datatype` is still `DeferredType`

Does PSyclone not check if an access to member of a structure is valid Fortran?

**Comment by arporter** [comment]
Ah, I hadn't appreciated we were looking at a member of a structure here. In theory, that should work as long as the type of `r2d_field` can be resolved. Is it being explicitly imported from a module?

**Comment by arporter** [comment]
I've just hit this again when trying to do inlining for LFRic. I've created an updated version of my original branch that extends InlineTrans.validate() to check partial_datatype if available: `2216_inline_partial_type`

**Comment by sergisiso** [comment]
I hit this issue with GPU offloading, we could recover the InlineKern concept or maybe lower the whole outer loop instead of just the kernel?

**Comment by sergisiso** [comment]
This is blocking GPU performance, so I have been evaluating different solutions:

- The lowering at transformation time (as we currently do in one GOcean example). This is way more complex in Lfric because there is lots of logic during lowering of other nodes (Loops, halos) that look for "kernels" and their metadata. Without this Lfric generates wrong results. Making this work would need a large reconsideration of the whole workflows.

- Having Kern subclass call or having the virtual arguments. This really doesn't solve the problem, the issue is preserving the metadata needed by other nodes during lowering.

- Re-purposing the InlinedKernel to keep be a node generated while inlining that keeps the metadata for queries by other nodes while lowering but insert the inlinded body as children. This has 2 problems: 1) currently it is not as simple as copying the metadata because it has many links that become broken and many assumptions in other places about the expected Invoke structure, 2) further modifying the body we could make changes that contradict the metadata. And the final code generation would be wrong.

- Adding a deferred-inline flag. Then during the lowering, after the metadata is not needed anymore and the Kern has been converted to a call, apply InlineTrans. This preserves the current workflow, doesn' copy/break the metadata, and minimises the possibilities of chaning code that contradicts the metadata (you still can, also now, by modifying the kernel module-inlined body, but it feels more constrained because you can not easily move things from inside the body to outside).

I am going to attempt the last option, it is by far the smaller change and is safe with the current workflow. I'm not convinced that it is the long-term solution (I was against the defered-module-inline flag that we had before), but it will allow us to check if InlineTrans supports current lfric code (I already see its not, e.g. complains about different precisions for unresolved i_def symbols) and validate the GPU performance behaves as we expect (by moving kernel local arrays allocations outside the loop).

In another PR I plan to remove metadata links, and make it an inmutable nested dataclass. So I don't discard the 3rd option yet for the future.

@arporter @LonelyCat124 Any thoughts?

**Comment by arporter** [comment]
Thanks Sergi. Of those options I would say number 3 feels like our best bet in the long term. I say this because if we defer it to the lowering stage, we can't then do further transformations - e.g. fusing in the PSy layer. (Although whether that will buy us anything with typical, large kernels is debatable.)  That said, in the spirit of trying things as soon as possible, I think trying the last option seems good.

**Comment by sergisiso** [comment]
I agree the 3 seems better long term and that fusing is going to be important, but not necessarely that this will be limited by 4.

Fusing in lfric just restructures the nodes to:
```
LFRicLoop
    LFRicKern
    LFRicKern
```
which does not impact the deferend-inline.

Another transformations that we could need "add atomics instead of colouring" also works with defered-inline.

But I agree for more advanced things the defered-inline is a problem.
